### PR TITLE
Group by keyword

### PR DESCRIPTION
**Pull request by mrkaiser**
* Added feature to group by keyword in the step glossary. `-T` is the argument
* Added an associated test

This allows people to know what steps are used by what keywords

### CODE REVIEW

**Review by bradsbrown — CHANGES_REQUESTED** [review]
@mrkaiser Thanks for the PR! I think this is a good concept, there are just a few tweaks to look at.

**Comment by bradsbrown** [comment]
@mrkaiser looks like the checks had some failures as well — be sure to review and correct those as you prepare new commits.

**Review by bradsbrown — APPROVED** [review]

**Comment by bradsbrown** [comment]
@mrkaiser Yep, looks great! I've added requests to get an additional review or two before merge.

File: sphinx_gherkindoc/cli.py
Comment (dgou) [inline sphinx_gherkindoc/cli.py:150]: First quick scan of this pr, this new parameter is a keyword parameter, so please pass it as such so that we maintain the flexibility to add other optional keyword parameters in the future and don't have to worry about the positional->keyword mappings being correct, thanks!

**Review by dgou — COMMENTED** [review]
One thing jumped out at me, so left an immediate comment, but I am planning a deeper review after work tonight or sometime this weekend.
Thanks, this is a cool new feature!

**Review by brolewis — APPROVED** [review]
Approving pending @dgou's final sign off

File: sphinx_gherkindoc/cli.py
Comment (dgou) [inline sphinx_gherkindoc/cli.py:150]: I like this much better, a lot more readable, thank you!

File: sphinx_gherkindoc/cli.py
Comment (dgou) [inline sphinx_gherkindoc/cli.py:217]: If we ever want to group by other mechanisms this might be obsolete.
I was going to suggest changing this to a choice list, but since I can't think of how else I would group it, that would be overkill.
Just jotting down thoughts, no change requested.

File: sphinx_gherkindoc/glossary.py
Comment (dgou) [inline sphinx_gherkindoc/glossary.py:73]: I had to stare at this for a while to work out the logic.
Conflicting interests:
1. Keeping the `if` in a postive form (`if x` is better than `if not x`)
2. Keeping the shorter code block as the `then` clause (not done here).

I was going to suggest preferring option 2 over option 1 as being net-simpler IMHO, as I was rewriting the code to be an example I kept stumbling over a deeper issue of how the code was refactored.
Boolean parameters are generally a code smell, two functions hiding together inside of one.
There are worse examples "in the wild", but it is bothering me.
HOWEVER, I am willing to say let's get this released and do a separate cleanup PR to refactor further since we now have some tests to make sure things still work.
I have some specific ideas about this, but I wanted to raise this so the author could also think about how to restructure it too.

**Review by dgou — APPROVED** [review]
LGTM. I think there are additional code improvements that could be made, but let's get this landed and sort that out in the next PR .
This is the red/GREEN/refactor PR, and it will be simpler if the red/green/REFACTOR was a separate PR. :-)

**Comment by dgou** [comment]
Thanks for the enhacement @mrkaiser 
Let me know if you want to take a shot at the refactoring.
I can let you know my thoughts, but I would like to hear yours before having any undue influence explaining mine.

**Comment by mrkaiser** [comment]
@dgou 

Hi! 

Thank you for all your thoughtful comments! 

Couple of thoughts:
* If there were multiple ways, I would prefer to have enums bound to functions that I could pass in. If a second use case comes through. I think that makes sense, so that a single "group by" could just take a group by  func. Personally I'm a fan of higher kinded functions and try to utilize them when possible. 
* I actually have tried to split it out. The key issue though is the `if group_by`. Eventually that if statement lives _somewhere_, even if it's early when doing the argparse.  Now you could continue and split it into two functions, say `make_steps_plain` and `make_steps_group_by`. But you'd have to reasonably make an additional helper because there is shared functionality in the preamble and the actual step building. The total code ends up a bit longer about 10 or so lines (which is about 10% of the total lines in the file). Now I have moved the main step building, and am sharing that code between both non and grouped by. Overall I understand the fear of bools. If they product _vastly_ different functions and you're hiding that with a "simple boolean", that is probably going to be hard to reason about later. I'm not entirely sure that's happening here, but would love to discuss further.
* The original if above the function for glossary writing was 
```
if not step_glossary
```
My thought process here was add to the if with the data structure I needed. Now we could collapse to 1 single data structure and then just get the flattened first. But that may be more logically complicated than what's here. But am happy to explore other ideas as well.

**Comment by dgou** [comment]
@mrkaiser Thanks for the response. It's been long enough now that I will have to sit with the code and get my brain to where it was from my previous comments. Will try to do that tonight, if not the next few nights.
Thanks!

**Comment by mrkaiser** [comment]
No rush at all. I won't be able to look until this weekend anyways. Appreciate any thoughts, really open to suggestions here. Overall I agree that booleans can be smells and try to avoid them. In this case though I just wanted to make sure I didn't over index on that and end up making something that may end up _harder_ to reason about.

**Comment by dgou** [comment]
Yeah, I thought I would get to it this week, but it looks like it'll be the weekend for me too.

**Pull request merged**
